File: ios-native/Sources/SKonnection/Models/CoffeeGame.swift
import Foundation

// 커피 내기 게임(커피룰렛 · 손가락룰렛) 공유 상태.
//
// 화면을 스트리밍하지 않는다. 주최자가 **seed 와 시작 시각**만 올리고, 각 기기가
// 같은 seed 로 똑같은 애니메이션을 스스로 그린다. 그래서
//   - 실시간 중계(모두 같은 걸 같은 순간에 본다)와
//   - 다시보기(나중에 같은 seed 로 재생)
// 가 같은 데이터 하나로 해결된다. 프레임 전송도, 영상 업로드도 없다.
//
// 저장은 신규 테이블 대신 기존 app_config(key, value jsonb)를 쓴다 —
// 프로덕 스키마를 건드리지 않기 위함. 키는 `coffee_game:<모임id>` 한 건.

/// 어떤 게임인지.
enum CoffeeGameKind: String, Codable, CaseIterable {
    case wheel = "커피룰렛"
    case finger = "손가락룰렛"

    var systemImage: String {
        switch self {
        case .wheel: return "arrow.triangle.2.circlepath"
        case .finger: return "hand.point.up.left.fill"
        }
    }
    var blurb: String {
        switch self {
        case .wheel: return "이름 칸으로 나뉜 원판을 돌립니다"
        case .finger: return "다 같이 한 폰에 손가락을 올립니다"
        }
    }
}

/// 게임 진행 단계. 관전자는 이 값으로 무엇을 보여줄지 정한다.
enum CoffeeGamePhase: String, Codable {
    case ready      // 주최자가 게임을 열었고 아직 안 돌림 — 관전자에겐 "곧 시작합니다"
    case spinning   // 돌아가는 중 — seed·startedAt 으로 각자 렌더
    case done       // 끝. winner 확정
}

struct CoffeeParticipant: Codable, Identifiable, Equatable {
    let name: String
    let photoURL: String
    var id: String { name }
}

struct CoffeeGame: Codable, Equatable {
    let gatheringId: String
    let kind: CoffeeGameKind
    var phase: CoffeeGamePhase
    /// 결정론적 재생용 난수 씨앗. spinning 이 되는 순간 정해진다.
    var seed: UInt64
    /// 돌릴 때 확정된 참가자(손가락룰렛은 그 순간 화면에 손이 올라온 사람만).
    var participants: [CoffeeParticipant]
    var winner: String
    /// spinning 시작 시각(epoch 밀리초). 늦게 들어온 관전자가 애니메이션 중간부터 맞춰 본다.
    ///
    /// MarketClock.iso 는 "yyyy-MM-dd'T'HH:mm" — **분 단위라 초가 없다.** 그걸 쓰면
    /// 기기마다 최대 60초까지 어긋나 동기화가 깨진다. 그래서 여기만 밀리초를 따로 쓴다.
    var startedAtMs: Int64
    var startedBy: String

    /// seed 로 정해지는 당첨자 인덱스. 모든 기기가 같은 답을 낸다.
    var winnerIndex: Int {
        guard !participants.isEmpty else { return 0 }
        var rng = SplitMix64(seed: seed)
        return Int(rng.next() % UInt64(participants.count))
    }
    var winnerName: String {
        participants.indices.contains(winnerIndex) ? participants[winnerIndex].name : ""
    }

    /// spinning 시작 후 흐른 시간(초). 관전자 화면 위치를 맞추는 기준.
    func elapsed(now: Date = Date()) -> Double {
        max(0, now.timeIntervalSince1970 - Double(startedAtMs) / 1000)
    }

    static func nowMs() -> Int64 { Int64(Date().timeIntervalSince1970 * 1000) }
}

/// 씨앗을 받는 작은 난수기(SplitMix64). Swift 기본 난수기는 씨앗을 못 넣어서
/// 기기마다 결과가 달라진다 — 결정론적 재생에는 쓸 수 없다.
struct SplitMix64: RandomNumberGenerator {
    private var state: UInt64
    init(seed: UInt64) { state = seed }
    mutating func next() -> UInt64 {
        state &+= 0x9E3779B97F4A7C15
        var z = state
        z = (z ^ (z >> 30)) &* 0xBF58476D1CE4E5B9
        z = (z ^ (z >> 27)) &* 0x94D049BB133111EB
        return z ^ (z >> 31)
    }
}

// MARK: - 저장소

@MainActor
final class CoffeeGameStore: ObservableObject {
    /// 모임 id → 현재 게임. 상세 화면이 열려 있는 동안만 폴링한다(배터리).
    @Published private(set) var games: [String: CoffeeGame] = [:]

    /// 이름 → 프로필 사진 URL. accounts 에서 한 번만 당겨 캐시한다.
    /// (사진은 사내 telinfo 서버라 사외망에서는 안 열릴 수 있다 — 그때는 이니셜로 뜬다.)
    @Published private(set) var photos: [String: String] = [:]

    private var pollTask: Task<Void, Never>?

    static func key(_ gatheringId: String) -> String { "coffee_game:\(gatheringId)" }

    func game(for gatheringId: String) -> CoffeeGame? { games[gatheringId] }

    /// 이름 목록을 사진이 붙은 참가자로 만든다.
    func participants(names: [String]) -> [CoffeeParticipant] {
        names.map { CoffeeParticipant(name: $0, photoURL: photos[$0] ?? "") }
    }

    func loadPhotosIfNeeded() async {
        guard photos.isEmpty, Supabase.isConfigured else { return }
        guard let rows = try? await Supabase.select(
            "accounts", query: "select=name,photo_url", as: AccountPhotoRow.self) else { return }
        photos = Dictionary(rows.compactMap { row -> (String, String)? in
            guard let n = row.name, let p = row.photo_url, !p.isEmpty else { return nil }
            return (n, p)
        }, uniquingKeysWith: { a, _ in a })
    }

    // MARK: 폴링(관전)

    /// 상세 화면이 열려 있는 동안 2초마다 서버 상태를 당겨온다.
    func startPolling(_ gatheringId: String) {
        stopPolling()
        pollTask = Task { @MainActor in
            while !Task.isCancelled {
                await refresh(gatheringId)
                try? await Task.sleep(nanoseconds: 2_000_000_000)
            }
        }
    }

    func stopPolling() {
        pollTask?.cancel()
        pollTask = nil
    }

    func refresh(_ gatheringId: String) async {
        guard Supabase.isConfigured else { return }
        let k = Self.key(gatheringId)
        guard let rows = try? await Supabase.select(
            "app_config", query: "key=eq.\(k)&select=value", as: CoffeeGameRow.self),
              let game = rows.first?.value else { return }
        games[gatheringId] = game
    }

    // MARK: 쓰기(주최자)

    /// 게임 화면을 열었다 — 관전자에게 "곧 시작합니다"를 알린다.
    func open(gatheringId: String, kind: CoffeeGameKind, host: String,
              participants: [CoffeeParticipant]) async {
        let game = CoffeeGame(gatheringId: gatheringId, kind: kind, phase: .ready, seed: 0,
                              participants: participants, winner: "",
                              startedAtMs: CoffeeGame.nowMs(), startedBy: host)
        games[gatheringId] = game
        await push(game)
    }

    /// 돌리기 시작 — 이 순간 seed 와 참가자가 확정된다.
    @discardableResult
    func spin(gatheringId: String, participants: [CoffeeParticipant]) async -> CoffeeGame? {
        guard var game = games[gatheringId] else { return nil }
        game.phase = .spinning
        game.seed = UInt64.random(in: 1...UInt64.max)
        game.participants = participants
        game.startedAtMs = CoffeeGame.nowMs()
        games[gatheringId] = game
        await push(game)
        return game
    }

    /// 연출이 끝났다 — 당첨자를 확정해 남긴다.
    func finish(gatheringId: String) async -> String? {
        guard var game = games[gatheringId], game.phase == .spinning else { return nil }
        game.phase = .done
        game.winner = game.winnerName
        games[gatheringId] = game
        await push(game)
        return game.winner
    }

    /// 다시 돌리기 위해 판을 비운다(아직 커피 담당이 확정되지 않았을 때만).
    func clear(gatheringId: String) async {
        games[gatheringId] = nil
        guard Supabase.isConfigured else { return }
        try? await Supabase.delete("app_config", query: "key=eq.\(Self.key(gatheringId))")
    }

    private func push(_ game: CoffeeGame) async {
        guard Supabase.isConfigured else { return }
        try? await Supabase.upsert("app_config",
                                   CoffeeGameUpsert(key: Self.key(game.gatheringId), value: game),
                                   onConflict: "key")
    }
}

// app_config.value 는 jsonb — 기존 NotifyValue 와 같이 중첩 객체로 그대로 넣고 뺀다.
struct CoffeeGameRow: Decodable { let value: CoffeeGame }
struct CoffeeGameUpsert: Encodable { let key: String; let value: CoffeeGame }
struct AccountPhotoRow: Decodable { let name: String?; let photo_url: String? }

File: android-native/app/src/main/java/com/hyubs/skonnection/data/CoffeeGame.kt
package com.hyubs.skonnection.data

import com.hyubs.skonnection.net.SupabaseClient
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.builtins.ListSerializer

// 커피 내기 게임(커피룰렛 · 손가락룰렛) — iOS CoffeeGame 이식.
// 화면을 스트리밍하지 않는다. 주최자가 seed·시작시각만 올리고, 각 기기가 같은 seed 로
// 똑같은 애니메이션을 스스로 그린다(실시간 중계 + 다시보기를 데이터 한 건으로).
// 저장은 기존 app_config(key, value jsonb) 재사용 — 키 `coffee_game:<모임id>`.
//
// **iOS/웹과 JSON 스키마·난수기가 반드시 일치**해야 iOS 주최 ↔ Android 관전에서
// 모두가 같은 당첨자를 본다(SplitMix64Test·camelCase 키).

@Serializable
enum class CoffeeGameKind {
    @SerialName("커피룰렛") WHEEL,
    @SerialName("손가락룰렛") FINGER;

    val label: String get() = if (this == WHEEL) "커피룰렛" else "손가락룰렛"
    val blurb: String get() = if (this == WHEEL) "이름 칸으로 나뉜 원판을 돌립니다" else "다 같이 한 폰에 손가락을 올립니다"
}

@Serializable
enum class CoffeeGamePhase {
    @SerialName("ready") READY,
    @SerialName("spinning") SPINNING,
    @SerialName("done") DONE,
}

@Serializable
data class CoffeeParticipant(
    val name: String,
    val photoURL: String = "",
)

@Serializable
data class CoffeeGame(
    val gatheringId: String,
    val kind: CoffeeGameKind,
    val phase: CoffeeGamePhase,
    val seed: ULong = 0uL,
    val participants: List<CoffeeParticipant> = emptyList(),
    val winner: String = "",
    val startedAtMs: Long = 0,
    val startedBy: String = "",
) {
    /** seed 로 정해지는 당첨자 인덱스 — 모든 기기가 같은 답을 낸다. */
    val winnerIndex: Int
        get() = if (participants.isEmpty()) 0 else (SplitMix64(seed).next() % participants.size.toULong()).toInt()

    val winnerName: String
        get() = participants.getOrNull(winnerIndex)?.name ?: ""

    /** spinning 시작 후 흐른 시간(초). 관전자 화면 위치를 맞추는 기준. */
    fun elapsed(nowMs: Long): Double = maxOf(0.0, (nowMs - startedAtMs) / 1000.0)

    companion object {
        fun key(gatheringId: String) = "coffee_game:$gatheringId"
    }
}

/**
 * 씨앗을 받는 결정론적 난수기(SplitMix64). iOS/웹과 **바이트 동일**해야 한다.
 * Kotlin ULong 연산은 자동 래핑되어 Swift `&+`/`&*` 와 같은 결과를 낸다.
 */
class SplitMix64(seed: ULong) {
    private var state: ULong = seed
    fun next(): ULong {
        state += 0x9E3779B97F4A7C15uL
        var z = state
        z = (z xor (z shr 30)) * 0xBF58476D1CE4E5B9uL
        z = (z xor (z shr 27)) * 0x94D049BB133111EBuL
        return z xor (z shr 31)
    }
}

/**
 * 두 게임의 연출 타임라인 — 시간만 넣으면 화면이 정해지는 순수 계산.
 * 주최자·관전자가 각자 그려도 같은 그림이 나온다(동기화의 핵심). iOS CoffeeSpin 이식.
 */
object CoffeeSpin {
    // ── 커피룰렛(원판) ──
    const val WHEEL_FAST = 1.4
    const val WHEEL_SLOW = 3.2
    val wheelTotalDuration get() = WHEEL_FAST + WHEEL_SLOW
    private const val WHEEL_FAST_SPEED = 900.0 // deg/s

    fun mod360(x: Double): Double {
        val m = x % 360.0
        return if (m < 0) m + 360 else m
    }

    /** 원판이 최종적으로 멈춰야 하는 각도(당첨 칸이 위쪽 바늘에 오는 각). */
    fun wheelTotalRotation(winnerIndex: Int, count: Int): Double {
        if (count <= 0) return 0.0
        val delta = 360.0 / count
        val align = mod360(-(winnerIndex + 0.5) * delta)
        val fast = WHEEL_FAST_SPEED * WHEEL_FAST
        val base = fast + 360 * 3
        return base + mod360(align - mod360(base))
    }

    /** 경과 시간 t 에서의 회전각. */
    fun wheelRotation(t: Double, winnerIndex: Int, count: Int): Double {
        val total = wheelTotalRotation(winnerIndex, count)
        if (t <= 0) return 0.0
        if (t < WHEEL_FAST) return WHEEL_FAST_SPEED * t
        val fast = WHEEL_FAST_SPEED * WHEEL_FAST
        val p = minOf(1.0, (t - WHEEL_FAST) / WHEEL_SLOW)
        val eased = 1 - Math.pow(1 - p, 3.0) // easeOutCubic
        return fast + (total - fast) * eased
    }

    // ── 손가락룰렛(두구두구) ──
    const val FINGER_DURATION = 3.0

    private fun fingerSteps(): List<Double> {
        val steps = ArrayList<Double>()
        var d = 0.075
        var sum = 0.0
        while (sum < FINGER_DURATION) {
            steps.add(d); sum += d; d *= 1.16
        }
        return steps
    }

    /** 경과 시간 t 에 포커스된 참가자 인덱스. 마지막 칸이 당첨자가 되도록 역산. */
    fun fingerFocus(t: Double, winnerIndex: Int, count: Int): Int {
        if (count <= 0) return 0
        val steps = fingerSteps()
        var acc = 0.0
        for (k in steps.indices) {
            acc += steps[k]
            if (t < acc) {
                val fromEnd = steps.size - 1 - k
                return ((winnerIndex - fromEnd) % count + count) % count
            }
        }
        return winnerIndex
    }
}

// ── app_config I/O ──
@Serializable
private data class CoffeeGameRow(val value: CoffeeGame)

@Serializable
private data class CoffeeGameUpsert(val key: String, val value: CoffeeGame)

@Serializable
private data class AccountPhotoRow(val name: String? = null, @SerialName("photo_url") val photoUrl: String? = null)

/** 커피 게임 원격 상태(app_config) + 참가자 사진(accounts.photo_url). */
class CoffeeGameRepository(private val supabase: SupabaseClient) {
    suspend fun load(gatheringId: String): CoffeeGame? =
        supabase.select(
            "app_config", "key=eq.${CoffeeGame.key(gatheringId)}&select=value",
            ListSerializer(CoffeeGameRow.serializer()),
        ).firstOrNull()?.value

    suspend fun push(game: CoffeeGame) =
        supabase.upsert(
            "app_config", CoffeeGameUpsert(CoffeeGame.key(game.gatheringId), game),
            CoffeeGameUpsert.serializer(), onConflict = "key",
        )

    suspend fun clear(gatheringId: String) =
        supabase.delete("app_config", "key=eq.${CoffeeGame.key(gatheringId)}")

    /** 이름 → 프로필 사진 URL. 사외망에선 안 열릴 수 있어 화면은 이니셜 폴백. */
    suspend fun loadPhotos(): Map<String, String> =
        supabase.select("accounts", "select=name,photo_url", ListSerializer(AccountPhotoRow.serializer()))
            .mapNotNull { r -> val n = r.name; val p = r.photoUrl; if (!n.isNullOrBlank() && !p.isNullOrBlank()) n to p else null }
            .toMap()
}
